x/interchainqueries/keeper/keeper.go

```diff
@@ -334,3 +379,55 @@ func (k Keeper) MustPayOutDeposit(ctx sdk.Context, deposit sdk.Coins, sender sdk
 		panic(err.Error())
 	}
 }
+
+// GetTxQueriesToRemove retrieves the list of TX queries registered to be removed. Returns a slice
+// with no more than limit entities or all entities if limit is 0.
+func (k Keeper) GetTxQueriesToRemove(ctx sdk.Context, limit uint64) []uint64 {
+	prefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.TxQueryToRemoveKey)
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+	ids := make([]uint64, 0, 100)
+	for ; iterator.Valid(); iterator.Next() {
+		ids = append(ids, sdk.BigEndianToUint64(iterator.Key()))
+		if limit != 0 && uint64(len(ids)) >= limit {
+			return ids
+		}
+	}
+	if len(ids) == 0 {
+		return nil
+	}
+	return ids
+}
+
+// calculateTxQueryRemoval creates a TxQueryToRemove populated with the data relative to the query
+// with the given queryID. The result TxQueryToRemove contains up to the limit tx hashes. If the
+// limit is 0, it retrieves all the hashes for the given query.
+func (k Keeper) calculateTxQueryRemoval(ctx sdk.Context, queryID uint64, limit uint64) *TxQueryToRemove {
+	prefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.GetSubmittedTransactionIDForQueryKeyPrefix(queryID))
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	result := &TxQueryToRemove{ID: queryID, Hashes: make([][]byte, 0, limit)}
+	for ; iterator.Valid(); iterator.Next() {
+		result.Hashes = append(result.Hashes, iterator.Key())
+		if limit != 0 && uint64(len(result.Hashes)) >= limit {
+			result.CompleteRemoval = !iterator.Valid()
+			return result
+		}
+	}
+	result.CompleteRemoval = true
+	return result
+}
+
+// TxQueryToRemove contains data related to a single query listed for removal and needed in the
+// removal process.
+type TxQueryToRemove struct {
```

Comment:
I don't really like having all the structs at the top of the file because of two reasons:

1. different structs have different degrees of importance to the reader. say, `Keeper` is the main structure in the `keeper.go` file and having it at the top is crucial because the reader will most likely look for this struct after opening this file. on the other hand, `TxQueryToRemove` is mostly an auxiliary struct;

2. I like keeping such a structure across files (roughly):
- consts
- struct A constructor and def
- struct A methods
- struct B constructor and def
- struct B methods
- ...
- auxiliary functions
in this way structure definition integrity is kept and one can read the whole structure without any breaks and distractions to other stuff.

hope it makes sense to you